doc/LAYERS.org

```diff
@@ -0,0 +1,588 @@
+
+- the =auto-completion= layer is not enabled, in which case no =init= function
+  for =company= will be found, or
+- the =company= package is excluded either by the user or another layer
+
+As such, =python/post-init-company= is the /only/ safe place to put
+configuration related to =company= in Python mode.
+
+If the Python layer had defined an =init= function for =company=, that package
+would have been installed even if the =auto-completion= layer had been disabled,
+which is not what we want.
+
+* Layer tips and tricks
+
+** Cross-dependencies
+Spacemacs provides a couple of additional useful functions you can use to check
+whether other layers or packages are included.
+
+- =configuration-layer/layer-usedp= :: check if a layer is enabled
+- =configuration-layer/package-usedp= :: check if a package is or will be installed
+
+These are useful in some cases, but usually you can get the desired result just
+by using =post-init= functions.
+
+For layers that require another layers to be enabled, use the functions
+=configuration-layer/declare-layer= and =configuration-layer/declare-layers= to
+ensure that layers are enabled even if the user has not enabled them explicitly.
+Calls to these functions should go in the =config.el= file.
+
+** Use-package
+In the vast majority of cases, a package =init= function should do nothing but
+call to =use-package=. Again, in the vast majority of cases, all the
+configuration you need to do should be doable within the =:init= or =:config=
+blocks of such a call.
+
+What goes where? Since =:init= is executed before load and =:config= after,
+these rules of thumb apply.
+
+In =:config= should be
+- Anything that requires the package to be already loaded.
+- Anything that takes a long time to run, which would ruin startup performance.
+
+The =:init= block should contain setup for the entry points to the package. This
+includes keybindings, if the package should be loaded manually by the user, or
+hooks, if the package should be loaded upon some event. It is not unusual to
+have both!
+
+** Use-package hooks
+Spacemacs includes a macro for adding more code to the =:init= or =:config=
+blocks of a call to =use-package=, after the fact. This is useful for =pre-init=
+or =post-init= functions to "inject" code into the =use-package= call of the
+=init= function.
+
+#+begin_src emacs-lisp
+  (spacemacs|use-package-add-hook helm
+    :pre-init
+    ;; Code
+    :post-init
+    ;; Code
+    :pre-config
+    ;; Code
+    :post-config
+    ;; Code
+    )
+#+end_src
+
+Since a call to =use-package= may evaluate the =:init= block immediately, any
+function that wants to inject code into this block must run =before= the call to
+=use-package=. Further, since this call to =use-package= typically takes place
+in the =init-<package>= function, calls to =spacemacs|use-package-add-hook=
+typically happen in the =pre-init-<package>= functions, and not in
+=post-init-<package>=.
+
```

Comment:
You can add that, as a rule of thumb, putting use-package hooks in `pre-init` is pretty safe so it should be the default choice when hooking stuff to use-package.